Sources/NIOPosix/System.swift

```diff
@@ -602,15 +602,17 @@ internal enum Posix {
 
     @inline(never)
     internal static func read(descriptor: CInt, pointer: UnsafeMutableRawPointer, size: size_t) throws -> IOResult<ssize_t> {
+        let cappedSize = size < Int32.max ? size : size_t(Int32.max)
```

Comment:
I don't think we should be capping this so low-level. This could lead to weird problems. IMHO we should make sure not to issue too-big `read` calls?

Reply:
Agreed, we should guard this in NonBlockingFileIO.

Reply:
Thanks for the input folks! I'm happy to make the change but could you possibly expand on the reasoning to inform future commits? It seemed to me that since this change was defensive against a low-level behaviour it made sense to 'protect' at as low a level as possible to be less intrusive and less likely that someone accidentally sidesteps the protection. 

For example Is it more important that we allow people to workaround this if desired or that there is higher visibility for this behaviour change?

Reply:
The syscall layer is not API: it's an implementation detail of NIO. That means the only people who call it are us. We tend to set ourselves the rule that we should solve problems at the location where we have the most information about what solution is optimal, and usually that's at as high a level as possible. It is not necessarily the case that it's safe to truncate arbitrary reads, but for NonBlockingFileIO we _know_ it's safe to do so.

Additionally, this helps with debugging. The syscall layer in NIO is essentially a direct translation of C syscalls into Swift. It does minimal data transformations. That means you don't have to look into it to see if it screwed with your data: you can assume that whatever you passed to it made it to the OS unchanged. That lets you avoid needing to have debugging sessions spend too much time in that layer of the code. This is a good principle to persist.

Reply:
Thanks Cory!